File: mobile_apps/job_automation_agent/backend/services/gemini_client.py
"""
Gemini API client for LLM interactions.
"""
import json
import logging
from typing import Optional, Dict, Any, List, Generator
from datetime import datetime
import google.generativeai as genai
from config import get_settings

logger = logging.getLogger(__name__)
settings = get_settings()


class CostTracker:
    """Track API costs for budget management."""
    
    # Gemini 1.5 Flash pricing (per 1M tokens)
    INPUT_COST_PER_MILLION = 0.075  # $0.075 per 1M input tokens
    OUTPUT_COST_PER_MILLION = 0.30  # $0.30 per 1M output tokens
    
    def __init__(self):
        self.daily_cost = 0.0
        self.total_calls = 0
        self.total_input_tokens = 0
        self.total_output_tokens = 0
        self.last_reset = datetime.utcnow().date()
    
    def _maybe_reset_daily(self):
        """Reset daily counters if it's a new day."""
        today = datetime.utcnow().date()
        if today > self.last_reset:
            self.daily_cost = 0.0
            self.last_reset = today
    
    def add_usage(self, input_tokens: int, output_tokens: int):
        """Record token usage and calculate cost."""
        self._maybe_reset_daily()
        
        input_cost = (input_tokens / 1_000_000) * self.INPUT_COST_PER_MILLION
        output_cost = (output_tokens / 1_000_000) * self.OUTPUT_COST_PER_MILLION
        cost = input_cost + output_cost
        
        self.daily_cost += cost
        self.total_calls += 1
        self.total_input_tokens += input_tokens
        self.total_output_tokens += output_tokens
        
        return cost
    
    def can_make_call(self) -> bool:
        """Check if we're within budget for another call."""
        self._maybe_reset_daily()
        return self.daily_cost < settings.max_daily_api_cost
    
    def get_stats(self) -> Dict[str, Any]:
        """Get current cost statistics."""
        return {
            "daily_cost": round(self.daily_cost, 4),
            "daily_budget": settings.max_daily_api_cost,
            "budget_remaining": round(settings.max_daily_api_cost - self.daily_cost, 4),
            "total_calls": self.total_calls,
            "total_input_tokens": self.total_input_tokens,
            "total_output_tokens": self.total_output_tokens
        }


class GeminiClient:
    """Client for interacting with Google Gemini API."""
    
    def __init__(self):
        if not settings.gemini_api_key:
            raise ValueError("GEMINI_API_KEY not configured")
        
        genai.configure(api_key=settings.gemini_api_key)
        self.model = genai.GenerativeModel(settings.gemini_model)
        self.cost_tracker = CostTracker()
        
        logger.info(f"Gemini client initialized with model: {settings.gemini_model}")
    
    def generate(
        self,
        prompt: str,
        system_prompt: Optional[str] = None,
        temperature: float = 0.7,
        max_tokens: int = 4096,
        json_mode: bool = False
    ) -> Dict[str, Any]:
        """
        Generate a response from Gemini.
        
        Returns:
            Dict with 'content', 'input_tokens', 'output_tokens', 'cost'
        """
        if not self.cost_tracker.can_make_call():
            raise RuntimeError("Daily API budget exceeded")
        
        try:
            # Build the full prompt
            full_prompt = ""
            if system_prompt:
                full_prompt = f"SYSTEM: {system_prompt}\n\nUSER: {prompt}"
            else:
                full_prompt = prompt
            
            # Configure generation
            generation_config = genai.GenerationConfig(
                temperature=temperature,
                max_output_tokens=max_tokens,
            )
            
            if json_mode:
                generation_config.response_mime_type = "application/json"
            
            # Generate response
            response = self.model.generate_content(
                full_prompt,
                generation_config=generation_config
            )
            
            # Extract token counts
            input_tokens = response.usage_metadata.prompt_token_count if response.usage_metadata else 0
            output_tokens = response.usage_metadata.candidates_token_count if response.usage_metadata else 0
            
            # Track cost
            cost = self.cost_tracker.add_usage(input_tokens, output_tokens)
            
            return {
                "content": response.text,
                "input_tokens": input_tokens,
                "output_tokens": output_tokens,
                "cost": cost
            }
            
        except Exception as e:
            logger.error(f"Gemini API error: {e}")
            raise
    
    def generate_json(
        self,
        prompt: str,
        system_prompt: Optional[str] = None,
        temperature: float = 0.3
    ) -> Dict[str, Any]:
        """Generate and parse JSON response."""
        result = self.generate(
            prompt=prompt,
            system_prompt=system_prompt,
            temperature=temperature,
            json_mode=True
        )
        
        try:
            parsed = json.loads(result["content"])
            result["parsed"] = parsed
            return result
        except json.JSONDecodeError as e:
            logger.error(f"Failed to parse JSON response: {e}")
            logger.debug(f"Raw content: {result['content']}")
            result["parsed"] = None
            result["parse_error"] = str(e)
            return result
    
    def analyze_job_fitment(
        self,
        profile_json: Dict[str, Any],
        job_description: str
    ) -> Dict[str, Any]:
        """
        Analyze job fitment using structured prompt.
        
        Returns:
            Dict with match_score, justification, strengths, weaknesses
        """
        system_prompt = """You are an expert career-matching AI. Your task is to analyze a user's profile against a job description and return a JSON object with a match score and a detailed analysis.

Be thorough but fair in your analysis. Consider:
- Skills alignment (both technical and soft skills)
- Years and type of experience
- Project relevance
- Industry knowledge
- Location and remote work compatibility"""

        user_prompt = f"""My Profile:
---
{json.dumps(profile_json, indent=2)}
---

Job Description:
---
{job_description}
---

Analyze my profile against the job description. Score the match from 0 to 100 based on:
1. Technical skills alignment (40%)
2. Experience relevance (30%)
3. Project/domain expertise (20%)
4. Cultural/soft skills fit (10%)

Provide a brief justification for your score, listing my strengths and weaknesses for this role.

Return ONLY a JSON object in the following format:
{{
  "match_score": <number>,
  "justification": {{
    "summary": "<brief overall assessment>",
    "strengths": ["<strength 1>", "<strength 2>", ...],
    "weaknesses": ["<weakness 1>", "<weakness 2>", ...],
    "missing_skills": ["<skill 1>", "<skill 2>", ...],
    "recommendations": ["<recommendation 1>", ...]
  }},
  "breakdown": {{
    "technical_skills": <0-100>,
    "experience": <0-100>,
    "projects": <0-100>,
    "soft_skills": <0-100>
  }}
}}"""

        result = self.generate_json(
            prompt=user_prompt,
            system_prompt=system_prompt,
            temperature=0.2
        )
        
        return result
    
    def classify_email(self, email_subject: str, email_body: str) -> Dict[str, Any]:
        """
        Classify an email related to job applications.
        """
        system_prompt = """You are an AI assistant that classifies job application-related emails.

Categories:
- APPLICATION_CONFIRMATION: Acknowledgment that application was received
- INTERVIEW_REQUEST: Request to schedule an interview
- REJECTION: The application was not successful
- OFFER: A job offer
- FOLLOWUP_REQUEST: Request for additional information
- ASSESSMENT_REQUEST: Request to complete an assessment/test
- GENERAL: Other job-related communication
- SPAM: Not related to job applications"""

        user_prompt = f"""Classify this email:

Subject: {email_subject}

Body:
{email_body[:2000]}  # Truncate for token efficiency

Return a JSON object:
{{
  "classification": "<category>",
  "confidence": <0.0-1.0>,
  "extracted_info": {{
    "company_name": "<if mentioned>",
    "job_title": "<if mentioned>",
    "interview_date": "<if scheduling>",
    "action_required": "<what user needs to do, if anything>",
    "deadline": "<any deadline mentioned>"
  }}
}}"""

        return self.generate_json(
            prompt=user_prompt,
            system_prompt=system_prompt,
            temperature=0.1
        )
    
    def customize_resume(
        self,
        profile_json: Dict[str, Any],
        job_description: str
    ) -> Dict[str, Any]:
        """
        Generate ATS-optimized resume content.
        """
        system_prompt = """You are an expert resume writer who specializes in optimizing resumes for Applicant Tracking Systems (ATS).

Your goal is to:
1. Identify key keywords from the job description
2. Rephrase experience and skills to match the job requirements
3. Highlight the most relevant projects and achievements
4. Ensure maximum ATS compatibility while maintaining authenticity"""

        user_prompt = f"""Profile:
{json.dumps(profile_json, indent=2)}

Job Description:
{job_description}

Create an ATS-optimized version of this profile. Return JSON:
{{
  "optimized_summary": "<new professional summary>",
  "highlighted_skills": ["<most relevant skill 1>", ...],
  "experience_bullets": [
    {{
      "company": "<company>",
      "role": "<role>",
      "bullets": ["<optimized bullet 1>", ...]
    }}
  ],
  "keywords_added": ["<keyword 1>", ...],
  "ats_score_estimate": <0-100>
}}"""

        return self.generate_json(
            prompt=user_prompt,
            system_prompt=system_prompt,
            temperature=0.4
        )
    
    def generate_interview_prep(
        self,
        profile_json: Dict[str, Any],
        job_description: str,
        company_info: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Generate comprehensive interview preparation document.
        """
        system_prompt = """You are an expert interview coach with deep knowledge of technical and behavioral interviews.

Create a comprehensive prep document that helps the candidate succeed."""

        company_context = f"\n\nCompany Information:\n{company_info}" if company_info else ""

        user_prompt = f"""Candidate Profile:
{json.dumps(profile_json, indent=2)}

Job Description:
{job_description}{company_context}

Generate interview preparation materials. Return JSON:
{{
  "company_research": {{
    "key_facts": ["<fact 1>", ...],
    "recent_news": ["<news item, if known>", ...],
    "culture_notes": "<insights about company culture>"
  }},
  "role_analysis": {{
    "key_responsibilities": ["<responsibility 1>", ...],
    "likely_challenges": ["<challenge 1>", ...],
    "success_metrics": ["<how success is measured>", ...]
  }},
  "behavioral_questions": [
    {{
      "question": "<question>",
      "why_asked": "<why they ask this>",
      "suggested_story": "<which experience to use>",
      "star_framework": {{
        "situation": "<brief situation>",
        "task": "<your task>",
        "action": "<what you did>",
        "result": "<outcome>"
      }}
    }}
  ],
  "technical_questions": [
    {{
      "question": "<technical question>",
      "topic": "<area being tested>",
      "preparation_tip": "<how to prepare>"
    }}
  ],
  "questions_to_ask": [
    {{
      "question": "<question for interviewer>",
      "shows_interest_in": "<what this demonstrates>"
    }}
  ],
  "talking_points": ["<key point to mention>", ...],
  "weakness_mitigation": {{
    "<weakness>": "<how to address if asked>"
  }}
}}"""

        return self.generate_json(
            prompt=user_prompt,
            system_prompt=system_prompt,
            temperature=0.5
        )
    
    def analyze_skill_gaps(
        self,
        profile_json: Dict[str, Any],
        missed_skills: List[str]
    ) -> Dict[str, Any]:
        """
        Analyze skill gaps and suggest learning resources.
        """
        system_prompt = """You are a career development advisor who specializes in helping professionals upskill efficiently.

Provide practical, actionable recommendations for learning new skills."""

        user_prompt = f"""Current Profile:
{json.dumps(profile_json, indent=2)}

Skills I'm missing for target jobs:
{json.dumps(missed_skills)}

For each missing skill, provide learning recommendations. Return JSON:
{{
  "skill_analysis": [
    {{
      "skill": "<skill name>",
      "importance": "<high/medium/low>",
      "time_to_competency": "<estimated time>",
      "learning_path": {{
        "beginner_resources": ["<resource 1>", ...],
        "intermediate_resources": ["<resource 1>", ...],
        "hands_on_projects": ["<project idea 1>", ...]
      }},
      "certification_options": ["<cert 1>", ...],
      "quick_wins": "<what can show progress quickly>"
    }}
  ],
  "priority_order": ["<skill to learn first>", "<skill 2>", ...],
  "synergies": "<skills that can be learned together>"
}}"""

        return self.generate_json(
            prompt=user_prompt,
            system_prompt=system_prompt,
            temperature=0.4
        )
    
    def chat(self, message: str, context: Optional[str] = None) -> Dict[str, Any]:
        """
        General chat for user queries about their job search.
        """
        system_prompt = """You are a helpful job search assistant. You have access to the user's profile, job applications, and search history.

Be concise, encouraging, and actionable. If asked about status, provide specific numbers and recommendations."""

        prompt = message
        if context:
            prompt = f"Context:\n{context}\n\nUser: {message}"
        
        return self.generate(
            prompt=prompt,
            system_prompt=system_prompt,
            temperature=0.7
        )
    
    def get_cost_stats(self) -> Dict[str, Any]:
        """Get current API cost statistics."""
        return self.cost_tracker.get_stats()


# Singleton instance
_client: Optional[GeminiClient] = None


def get_gemini_client() -> GeminiClient:
    """Get or create Gemini client singleton."""
    global _client
    if _client is None:
        _client = GeminiClient()
    return _client

File: mobile_apps/job_automation_standalone/app/src/main/java/com/vignesh/jobautomation/ai/GeminiClient.kt
package com.vignesh.jobautomation.ai

import android.util.Log
import com.google.ai.client.generativeai.GenerativeModel
import com.google.ai.client.generativeai.type.GenerateContentResponse
import com.google.ai.client.generativeai.type.generationConfig
import com.squareup.moshi.JsonAdapter
import com.squareup.moshi.Moshi
import com.squareup.moshi.Types
import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
import com.vignesh.jobautomation.BuildConfig
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * Gemini AI Client for job automation tasks.
 * Handles all AI-powered analyses including job matching, resume customization, and interview prep.
 */
class GeminiClient private constructor(apiKey: String) {
    
    private val model = GenerativeModel(
        modelName = "gemini-1.5-flash",
        apiKey = apiKey,
        generationConfig = generationConfig {
            temperature = 0.3f
            maxOutputTokens = 4096
        }
    )
    
    private val moshi = Moshi.Builder()
        .addLast(KotlinJsonAdapterFactory())
        .build()
    
    companion object {
        private const val TAG = "GeminiClient"
        
        @Volatile
        private var INSTANCE: GeminiClient? = null
        
        fun getInstance(apiKey: String = BuildConfig.GEMINI_API_KEY): GeminiClient {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: GeminiClient(apiKey).also { INSTANCE = it }
            }
        }
    }
    
    /**
     * Analyze job fitment against user profile.
     * Returns match score (0-100) with justification.
     */
    suspend fun analyzeJobFitment(
        profileSummary: String,
        skills: Map<String, List<String>>,
        experience: String,
        desiredRoles: List<String>,
        jobDescription: String
    ): JobAnalysisResult = withContext(Dispatchers.IO) {
        val prompt = buildString {
            appendLine("You are an expert job matching analyst. Analyze how well this candidate matches the job.")
            appendLine()
            appendLine("=== CANDIDATE PROFILE ===")
            appendLine("Summary: $profileSummary")
            appendLine()
            appendLine("Skills:")
            skills.forEach { (category, skillList) ->
                appendLine("- $category: ${skillList.joinToString(", ")}")
            }
            appendLine()
            appendLine("Experience: $experience")
            appendLine()
            appendLine("Desired Roles: ${desiredRoles.joinToString(", ")}")
            appendLine()
            appendLine("=== JOB DESCRIPTION ===")
            appendLine(jobDescription.take(3000))
            appendLine()
            appendLine("=== ANALYSIS REQUIRED ===")
            appendLine("Provide a JSON response with EXACTLY this structure:")
            appendLine("""
{
  "match_score": 75,
  "summary": "Brief 1-sentence summary of match",
  "strengths": ["strength 1", "strength 2", "strength 3"],
  "weaknesses": ["gap 1", "gap 2"],
  "key_requirements_met": ["requirement 1", "requirement 2"],
  "key_requirements_missing": ["missing 1"],
  "salary_range_estimate": {"min": 120000, "max": 180000},
  "recommendation": "APPLY" | "CONSIDER" | "SKIP",
  "next_steps": "What the candidate should do"
}
            """.trimIndent())
            appendLine()
            appendLine("IMPORTANT: Return ONLY valid JSON, no markdown, no extra text.")
        }
        
        try {
            val response = model.generateContent(prompt)
            val text = response.text?.trim() ?: throw Exception("Empty response")
            parseJobAnalysis(text)
        } catch (e: Exception) {
            Log.e(TAG, "Job analysis failed", e)
            JobAnalysisResult(
                matchScore = 0f,
                summary = "Analysis failed: ${e.message}",
                strengths = emptyList(),
                weaknesses = emptyList(),
                recommendation = "SKIP"
            )
        }
    }
    
    /**
     * Generate a customized resume for a specific job.
     */
    suspend fun customizeResume(
        profileJson: String,
        jobDescription: String
    ): ResumeCustomization = withContext(Dispatchers.IO) {
        val prompt = buildString {
            appendLine("You are an expert ATS resume optimizer. Customize this resume for the job.")
            appendLine()
            appendLine("=== CANDIDATE PROFILE (JSON) ===")
            appendLine(profileJson.take(4000))
            appendLine()
            appendLine("=== TARGET JOB ===")
            appendLine(jobDescription.take(2000))
            appendLine()
            appendLine("=== CUSTOMIZATION REQUIRED ===")
            appendLine("Provide a JSON response with EXACTLY this structure:")
            appendLine("""
{
  "optimized_summary": "ATS-optimized professional summary (2-3 sentences)",
  "highlighted_skills": ["skill1", "skill2", "skill3"],
  "experience_bullets": [
    {
      "company": "Company Name",
      "bullets": ["Achievement-focused bullet 1", "Bullet 2"]
    }
  ],
  "keywords_to_add": ["keyword1", "keyword2"],
  "ats_score_estimate": 85,
  "cover_letter_opening": "Compelling opening paragraph"
}
            """.trimIndent())
            appendLine()
            appendLine("IMPORTANT: Return ONLY valid JSON.")
        }
        
        try {
            val response = model.generateContent(prompt)
            val text = response.text?.trim() ?: throw Exception("Empty response")
            parseResumeCustomization(text)
        } catch (e: Exception) {
            Log.e(TAG, "Resume customization failed", e)
            ResumeCustomization(
                optimizedSummary = "",
                highlightedSkills = emptyList(),
                experienceBullets = emptyList(),
                keywordsToAdd = emptyList(),
                atsScoreEstimate = 0,
                coverLetterOpening = ""
            )
        }
    }
    
    /**
     * Generate interview preparation materials.
     */
    suspend fun generateInterviewPrep(
        jobTitle: String,
        companyName: String,
        jobDescription: String,
        profileSummary: String
    ): InterviewPrepResult = withContext(Dispatchers.IO) {
        val prompt = buildString {
            appendLine("You are an expert interview coach. Prepare the candidate for this interview.")
            appendLine()
            appendLine("=== JOB ===")
            appendLine("Title: $jobTitle at $companyName")
            appendLine("Description: ${jobDescription.take(2000)}")
            appendLine()
            appendLine("=== CANDIDATE ===")
            appendLine(profileSummary.take(1500))
            appendLine()
            appendLine("=== INTERVIEW PREP REQUIRED ===")
            appendLine("Provide a JSON response with EXACTLY this structure:")
            appendLine("""
{
  "company_research": {
    "key_facts": ["fact1", "fact2"],
    "recent_news": ["news1", "news2"],
    "culture_notes": "Company culture insights"
  },
  "likely_questions": [
    {
      "question": "Interview question",
      "type": "behavioral" | "technical" | "situational",
      "suggested_answer_points": ["point1", "point2"],
      "star_example": "Optional STAR format example"
    }
  ],
  "questions_to_ask": ["Question 1 to ask interviewer", "Question 2"],
  "technical_topics": ["Topic to review 1", "Topic 2"],
  "red_flags_to_avoid": ["Avoid saying X", "Don't forget Y"],
  "preparation_checklist": ["Research company", "Review portfolio", "Prepare questions"]
}
            """.trimIndent())
            appendLine()
            appendLine("IMPORTANT: Return ONLY valid JSON.")
        }
        
        try {
            val response = model.generateContent(prompt)
            val text = response.text?.trim() ?: throw Exception("Empty response")
            parseInterviewPrep(text)
        } catch (e: Exception) {
            Log.e(TAG, "Interview prep failed", e)
            InterviewPrepResult(
                companyResearch = CompanyResearch(emptyList(), emptyList(), ""),
                likelyQuestions = emptyList(),
                questionsToAsk = emptyList(),
                technicalTopics = emptyList(),
                redFlagsToAvoid = emptyList(),
                preparationChecklist = emptyList()
            )
        }
    }
    
    /**
     * Analyze skill gaps and provide improvement recommendations.
     */
    suspend fun analyzeWeaknesses(
        currentSkills: Map<String, List<String>>,
        targetRoles: List<String>,
        recentJobMisses: List<String>
    ): WeaknessAnalysis = withContext(Dispatchers.IO) {
        val prompt = buildString {
            appendLine("You are a career development advisor. Analyze skill gaps and provide improvement plan.")
            appendLine()
            appendLine("=== CURRENT SKILLS ===")
            currentSkills.forEach { (category, skills) ->
                appendLine("- $category: ${skills.joinToString(", ")}")
            }
            appendLine()
            appendLine("=== TARGET ROLES ===")
            appendLine(targetRoles.joinToString(", "))
            appendLine()
            appendLine("=== RECENT JOB WEAKNESSES IDENTIFIED ===")
            recentJobMisses.take(5).forEachIndexed { i, weakness ->
                appendLine("${i + 1}. $weakness")
            }
            appendLine()
            appendLine("=== ANALYSIS REQUIRED ===")
            appendLine("Provide a JSON response:")
            appendLine("""
{
  "critical_gaps": [
    {
      "skill": "Skill name",
      "importance": "HIGH" | "MEDIUM" | "LOW",
      "current_level": "Beginner/Intermediate/Advanced/None",
      "target_level": "What level is needed",
      "time_to_learn": "Estimated time"
    }
  ],
  "learning_resources": [
    {
      "skill": "Skill name",
      "resources": ["Resource 1", "Resource 2"],
      "certifications": ["Cert 1 if applicable"]
    }
  ],
  "quick_wins": ["Improvement that can be done quickly"],
  "long_term_plan": "6-month improvement roadmap",
  "portfolio_suggestions": ["Project idea to demonstrate skill"]
}
            """.trimIndent())
            appendLine()
            appendLine("IMPORTANT: Return ONLY valid JSON.")
        }
        
        try {
            val response = model.generateContent(prompt)
            val text = response.text?.trim() ?: throw Exception("Empty response")
            parseWeaknessAnalysis(text)
        } catch (e: Exception) {
            Log.e(TAG, "Weakness analysis failed", e)
            WeaknessAnalysis(
                criticalGaps = emptyList(),
                learningResources = emptyList(),
                quickWins = emptyList(),
                longTermPlan = "",
                portfolioSuggestions = emptyList()
            )
        }
    }
    
    /**
     * Chat with the AI about job search topics.
     */
    suspend fun chat(
        message: String,
        context: String? = null,
        chatHistory: List<Pair<String, String>> = emptyList()
    ): String = withContext(Dispatchers.IO) {
        val prompt = buildString {
            appendLine("You are a helpful job search assistant. Help the user with their job search.")
            appendLine()
            if (context != null) {
                appendLine("=== CONTEXT ===")
                appendLine(context.take(2000))
                appendLine()
            }
            if (chatHistory.isNotEmpty()) {
                appendLine("=== CONVERSATION HISTORY ===")
                chatHistory.takeLast(5).forEach { (role, content) ->
                    appendLine("$role: $content")
                }
                appendLine()
            }
            appendLine("=== USER MESSAGE ===")
            appendLine(message)
            appendLine()
            appendLine("Respond helpfully and concisely. If discussing jobs or applications, be specific and actionable.")
        }
        
        try {
            val response = model.generateContent(prompt)
            response.text?.trim() ?: "I couldn't generate a response. Please try again."
        } catch (e: Exception) {
            Log.e(TAG, "Chat failed", e)
            "Sorry, I encountered an error: ${e.message}"
        }
    }
    
    // ==================== JSON Parsing Helpers ====================
    
    private fun parseJobAnalysis(json: String): JobAnalysisResult {
        val cleanJson = cleanJsonResponse(json)
        return try {
            val type = Types.newParameterizedType(Map::class.java, String::class.java, Any::class.java)
            val adapter: JsonAdapter<Map<String, Any>> = moshi.adapter(type)
            val map = adapter.fromJson(cleanJson) ?: emptyMap()
            
            JobAnalysisResult(
                matchScore = (map["match_score"] as? Number)?.toFloat() ?: 0f,
                summary = map["summary"] as? String ?: "",
                strengths = (map["strengths"] as? List<*>)?.filterIsInstance<String>() ?: emptyList(),
                weaknesses = (map["weaknesses"] as? List<*>)?.filterIsInstance<String>() ?: emptyList(),
                keyRequirementsMet = (map["key_requirements_met"] as? List<*>)?.filterIsInstance<String>() ?: emptyList(),
                keyRequirementsMissing = (map["key_requirements_missing"] as? List<*>)?.filterIsInstance<String>() ?: emptyList(),
                salaryEstimate = (map["salary_range_estimate"] as? Map<*, *>)?.let {
                    SalaryEstimate((it["min"] as? Number)?.toInt() ?: 0, (it["max"] as? Number)?.toInt() ?: 0)
                },
                recommendation = map["recommendation"] as? String ?: "CONSIDER",
                nextSteps = map["next_steps"] as? String ?: ""
            )
        } catch (e: Exception) {
            Log.e(TAG, "Failed to parse job analysis: $cleanJson", e)
            JobAnalysisResult(matchScore = 0f, summary = "Parse error", strengths = emptyList(), weaknesses = emptyList(), recommendation = "SKIP")
        }
    }
    
    private fun parseResumeCustomization(json: String): ResumeCustomization {
        val cleanJson = cleanJsonResponse(json)
        return try {
            val type = Types.newParameterizedType(Map::class.java, String::class.java, Any::class.java)
            val adapter: JsonAdapter<Map<String, Any>> = moshi.adapter(type)
            val map = adapter.fromJson(cleanJson) ?: emptyMap()
            
            ResumeCustomization(
                optimizedSummary = map["optimized_summary"] as? String ?: "",
                highlightedSkills = (map["highlighted_skills"] as? List<*>)?.filterIsInstance<String>() ?: emptyList(),
                experienceBullets = (map["experience_bullets"] as? List<*>)?.mapNotNull { item ->
                    (item as? Map<*, *>)?.let { m ->
                        ExperienceBullet(
                            company = m["company"] as? String ?: "",
                            bullets = (m["bullets"] as? List<*>)?.filterIsInstance<String>() ?: emptyList()
                        )
                    }
                } ?: emptyList(),
                keywordsToAdd = (map["keywords_to_add"] as? List<*>)?.filterIsInstance<String>() ?: emptyList(),
                atsScoreEstimate = (map["ats_score_estimate"] as? Number)?.toInt() ?: 0,
                coverLetterOpening = map["cover_letter_opening"] as? String ?: ""
            )
        } catch (e: Exception) {
            Log.e(TAG, "Failed to parse resume customization", e)
            ResumeCustomization("", emptyList(), emptyList(), emptyList(), 0, "")
        }
    }
    
    private fun parseInterviewPrep(json: String): InterviewPrepResult {
        val cleanJson = cleanJsonResponse(json)
        return try {
            val type = Types.newParameterizedType(Map::class.java, String::class.java, Any::class.java)
            val adapter: JsonAdapter<Map<String, Any>> = moshi.adapter(type)
            val map = adapter.fromJson(cleanJson) ?: emptyMap()
            
            val companyResearchMap = map["company_research"] as? Map<*, *>
            val companyResearch = CompanyResearch(
                keyFacts = (companyResearchMap?.get("key_facts") as? List<*>)?.filterIsInstance<String>() ?: emptyList(),
                recentNews = (companyResearchMap?.get("recent_news") as? List<*>)?.filterIsInstance<String>() ?: emptyList(),
                cultureNotes = companyResearchMap?.get("culture_notes") as? String ?: ""
            )
            
            val questions = (map["likely_questions"] as? List<*>)?.mapNotNull { item ->
                (item as? Map<*, *>)?.let { q ->
                    InterviewQuestion(
                        question = q["question"] as? String ?: "",
                        type = q["type"] as? String ?: "general",
                        suggestedAnswerPoints = (q["suggested_answer_points"] as? List<*>)?.filterIsInstance<String>() ?: emptyList(),
                        starExample = q["star_example"] as? String
                    )
                }
            } ?: emptyList()
            
            InterviewPrepResult(
                companyResearch = companyResearch,
                likelyQuestions = questions,
                questionsToAsk = (map["questions_to_ask"] as? List<*>)?.filterIsInstance<String>() ?: emptyList(),
                technicalTopics = (map["technical_topics"] as? List<*>)?.filterIsInstance<String>() ?: emptyList(),
                redFlagsToAvoid = (map["red_flags_to_avoid"] as? List<*>)?.filterIsInstance<String>() ?: emptyList(),
                preparationChecklist = (map["preparation_checklist"] as? List<*>)?.filterIsInstance<String>() ?: emptyList()
            )
        } catch (e: Exception) {
            Log.e(TAG, "Failed to parse interview prep", e)
            InterviewPrepResult(CompanyResearch(emptyList(), emptyList(), ""), emptyList(), emptyList(), emptyList(), emptyList(), emptyList())
        }
    }
    
    private fun parseWeaknessAnalysis(json: String): WeaknessAnalysis {
        val cleanJson = cleanJsonResponse(json)
        return try {
            val type = Types.newParameterizedType(Map::class.java, String::class.java, Any::class.java)
            val adapter: JsonAdapter<Map<String, Any>> = moshi.adapter(type)
            val map = adapter.fromJson(cleanJson) ?: emptyMap()
            
            WeaknessAnalysis(
                criticalGaps = (map["critical_gaps"] as? List<*>)?.mapNotNull { item ->
                    (item as? Map<*, *>)?.let { g ->
                        SkillGap(
                            skill = g["skill"] as? String ?: "",
                            importance = g["importance"] as? String ?: "MEDIUM",
                            currentLevel = g["current_level"] as? String ?: "",
                            targetLevel = g["target_level"] as? String ?: "",
                            timeToLearn = g["time_to_learn"] as? String ?: ""
                        )
                    }
                } ?: emptyList(),
                learningResources = (map["learning_resources"] as? List<*>)?.mapNotNull { item ->
                    (item as? Map<*, *>)?.let { r ->
                        LearningResource(
                            skill = r["skill"] as? String ?: "",
                            resources = (r["resources"] as? List<*>)?.filterIsInstance<String>() ?: emptyList(),
                            certifications = (r["certifications"] as? List<*>)?.filterIsInstance<String>() ?: emptyList()
                        )
                    }
                } ?: emptyList(),
                quickWins = (map["quick_wins"] as? List<*>)?.filterIsInstance<String>() ?: emptyList(),
                longTermPlan = map["long_term_plan"] as? String ?: "",
                portfolioSuggestions = (map["portfolio_suggestions"] as? List<*>)?.filterIsInstance<String>() ?: emptyList()
            )
        } catch (e: Exception) {
            Log.e(TAG, "Failed to parse weakness analysis", e)
            WeaknessAnalysis(emptyList(), emptyList(), emptyList(), "", emptyList())
        }
    }
    
    private fun cleanJsonResponse(text: String): String {
        var clean = text.trim()
        // Remove markdown code blocks
        if (clean.startsWith("```json")) {
            clean = clean.removePrefix("```json")
        }
        if (clean.startsWith("```")) {
            clean = clean.removePrefix("```")
        }
        if (clean.endsWith("```")) {
            clean = clean.removeSuffix("```")
        }
        return clean.trim()
    }
}

// ==================== Data Classes ====================

data class JobAnalysisResult(
    val matchScore: Float,
    val summary: String,
    val strengths: List<String>,
    val weaknesses: List<String>,
    val keyRequirementsMet: List<String> = emptyList(),
    val keyRequirementsMissing: List<String> = emptyList(),
    val salaryEstimate: SalaryEstimate? = null,
    val recommendation: String, // "APPLY", "CONSIDER", "SKIP"
    val nextSteps: String = ""
)

data class SalaryEstimate(
    val min: Int,
    val max: Int
)

data class ResumeCustomization(
    val optimizedSummary: String,
    val highlightedSkills: List<String>,
    val experienceBullets: List<ExperienceBullet>,
    val keywordsToAdd: List<String>,
    val atsScoreEstimate: Int,
    val coverLetterOpening: String
)

data class ExperienceBullet(
    val company: String,
    val bullets: List<String>
)

data class InterviewPrepResult(
    val companyResearch: CompanyResearch,
    val likelyQuestions: List<InterviewQuestion>,
    val questionsToAsk: List<String>,
    val technicalTopics: List<String>,
    val redFlagsToAvoid: List<String>,
    val preparationChecklist: List<String>
)

data class CompanyResearch(
    val keyFacts: List<String>,
    val recentNews: List<String>,
    val cultureNotes: String
)

data class InterviewQuestion(
    val question: String,
    val type: String,
    val suggestedAnswerPoints: List<String>,
    val starExample: String?
)

data class WeaknessAnalysis(
    val criticalGaps: List<SkillGap>,
    val learningResources: List<LearningResource>,
    val quickWins: List<String>,
    val longTermPlan: String,
    val portfolioSuggestions: List<String>
)

data class SkillGap(
    val skill: String,
    val importance: String,
    val currentLevel: String,
    val targetLevel: String,
    val timeToLearn: String
)

data class LearningResource(
    val skill: String,
    val resources: List<String>,
    val certifications: List<String>
)
